### ISSUE TITLE
Browser support

### ISSUE THREAD

**Issue by mattdesl** [avajs/ava#24]
Trying to get this working in the browser, and then later [hihat](https://github.com/Jam3/hihat) for debugging/breakpoints/etc.

After browserifying the demo in the readme, I get the following:

`Uncaught ReferenceError: setImmediate is not defined`

[relevant line](https://github.com/sindresorhus/ava/blob/97add1fcec62d510d72f1d26e1ea5ca972c495c6/index.js#L65)

(weird cause I thought browserify would have shimmed this)

**Comment by sindresorhus** [comment]
Maybe open an issue on Browserify about shimming `setImmediate`?

**Comment by mattdesl** [comment]
Found a bit more discussion upstream: https://github.com/substack/insert-module-globals/pull/40

**Comment by sindresorhus** [comment]
I'll give it a week, but I doubt anything will happen there... 

I guess I can add https://github.com/sindresorhus/set-immediate-shim if that is the only thing blocking it from being used in the browser.

**Comment by Qix-** [comment]
Isn't it just a fancy way of doing `setTimeout(fn, 0)` or `Window.requestAnimationFrame`?

**Comment by sindresorhus** [comment]
@Qix- It's faster than `setTimeout`. http://www.nczonline.net/blog/2013/07/09/the-case-for-setimmediate/

**Comment by tomekwi** [comment]
@vdemedes thanks for the fix! The `setImmediate` error is busted.

But now I’ve got another bastard there:

```
Uncaught TypeError: Cannot read property 'write' of undefined
```

– coming from https://github.com/kevva/squeak/blob/v1.2.0/index.js#L144.

**Comment by vdemedes** [comment]
@tomekwi You're welcome!

I see what the problem is, will try to not forget and fix it!

**Comment by tomekwi** [comment]
Great! I can’t wait to check out this awesome trio: *hihat*, *ava* and *[quixote](http://npm.im/quixote)*! *quixote* doesn’t go too well with *tape*, because it’s actually an assertion library.

**Comment by Qix-** [comment]
Another cool tool, `quixote`. Thanks for the link :dancer:

**Comment by mattdesl** [comment]
In the browser `process.stdout` doesn't exist and browserify doesn't try to shim it. Tape uses the following to print in the browser:
https://github.com/substack/tape/blob/master/lib/default_stream.js

**Comment by Qix-** [comment]
Shimming `process.stdout` would be simple, assuming somewhere along the way you write a newline.

**Comment by danielepolencic** [comment]
same issue with https://github.com/kevva/squeak/blob/v1.2.0/index.js#L144 :(

**Comment by sindresorhus** [comment]
Browserify should already shim `process`, so sounds like either a Browserify bug or an issue with your build setup.

**Comment by danielepolencic** [comment]
Browserify shims `process`, but looks like it doesn't shim `process.stderr`.

I think that causes https://github.com/kevva/squeak/blob/v1.2.0/index.js#L26 to be undefined and line 144 to fail.

**Comment by sindresorhus** [comment]
https://github.com/defunctzombie/node-process/issues/35

I guess we can make `squeak` use `console.log`/`console.error` instead in the browser.

**Comment by danielepolencic** [comment]
I [tweaked the constructor](https://github.com/sindresorhus/ava/blob/5b6c18138de99d2d447859a8c7ec955203c8e084/index.js#L9) and injected the console in `squeak` like you suggested. 

This seems to fix the issue. I can run `ava` in `browserify` + `hihat`. :tada: :tada:

I get some errors about `process.exit` and `bluebird`, though.

ATM, I think it would be better to monkey patch `process.stderr` and `process.exit` rather than just injecting `console` in `squeak` - like you suggested in that link.

**Comment by sindresorhus** [comment]
Yeah. I guess we could replace the `logger.js` file with a custom `logger-browser.js` using the Browserify replace feature.

And `process.exit = window.close;`. https://github.com/Jam3/hihat#basic-examples

> The process will stay open until you call window.close() from the client code.

**Comment by sindresorhus** [comment]
Happy to recommend `hihat` in the readme when we get this working :)

**Comment by mattdesl** [comment]
Added a ticket about exit: https://github.com/Jam3/hihat/issues/27

Not sure about shimming stderr/stdout in browserify. Seems like browserify's `process` module could provide a polyfill, but all the stream stuff could add a lot of extra bytes to bundles for mostly no gain.

**Comment by mattdesl** [comment]
With that said; maybe a dumb polyfill could just write console.log and error synchronously, but not attempt to support any other aspect of the stdout/stderr streams.

**Comment by danielepolencic** [comment]
[This is the line](https://github.com/sindresorhus/ava/blob/5b6c18138de99d2d447859a8c7ec955203c8e084/index.js#L86) that fails with the `process.exit`.

As for [process.stderr](https://github.com/kevva/squeak/blob/v1.2.0/index.js#L26) I think the issue is not as straightforward as I thought.

`ava` *IS* browserify friendly (a part from `process.exit`), but `squeak` is not. In theory, the latter should be patched and not `ava`.

The issue with `squeak` is that even if I [patch the stream](https://github.com/sindresorhus/ava/blob/5b6c18138de99d2d447859a8c7ec955203c8e084/index.js#L9) to redirect to `console`, all output would be redirected to `console.log` and we lose all log types (i.e. error, warn, info, etc.). What it would be better is to patch `squeak` (or use another library) so that it falls back to console and still maintains all the right logging levels.

**Comment by sindresorhus** [comment]
> What it would be better is to patch squeak (or use another library)

Agreed.

> so that it falls back to console and still maintains all the right logging levels.

How would it do that? Prefix the console output with logging level?

**Comment by danielepolencic** [comment]
With the following (hacky) change in `index.js`:

```js
var log = new Squeak({separator: ' ', stream: {write: console.log.bind(console)}});
```

all the logs from type `success` and `error` are redirected to `console.log`.

Ideally, one would like to have the following mappings between `squeak` and `console`

Squeak Type | logging level
------------|--------------
info        | console.info
success     | console.log
error       | console.error
warn        | console.warn

[winston](https://github.com/winstonjs/winston) offers such functionality (and multiple transports too), but it's not browserify friendly.

I think the quickest win in this case would be to use `console` directly? 

as an example `log.error` could be emulated with:

```js
var logError = console.error.bind(console, figures.cross);
```

That would work in node and in the browser.

**Comment by sindresorhus** [comment]
Shouldn't be hard to patch `squeak` to use `console.log` when used in the browser though. I think that would be the best and less hacky solution.

**Comment by phuu** [comment]
Hey ava peeps — could someone please make issues for the work that needs doing on ava to make running in a browser possible (or document how to do it!) so I or someone else could work on it? Thanks!

**Comment by ariporad** [comment]
@sindresorhus: Do we want to start adding browser support? Or are we waiting on that?

If we're going ahead with it, I'd like to help (I'm going back to school soon, so I might not be as active). Would you want to just abandon multi-threaded-ness? Or should we run things in web workers? Or new tabs? iframes? Also, I presume you'd want to integrate with Karma?

**Comment by jamestalmage** [comment]
My vote is to wait. Having to support browsers will slow us way down. I think browser support should be just about the last thing before `1.0.0`.

**Comment by ariporad** [comment]
@jamestalmage: Agreed

**Comment by sindresorhus** [comment]
I have no idea anymore what it would take to make it run in the browser. Browser support comes with an immense amount of overhead, both in edge-cases and user support. I want for us to finish the core of AVA before committing to something like this. We already have more than enough overhead with Babel.

That being said, I don't see any reason we can't discuss potential solutions, and anyone is free to look into how to achieve this. It might be as simple as replacing `lib/fork.js` with a shim for the browser that just communicates directly with `lib/child-worker.js` (browserify supports file overrides for the browser), or it might require a lot more. I know this isn't what people were hoping for, but we have to prioritize.

**Comment by phuu** [comment]
Ok. It's disappointing but I totally understand — do what's right for this project :smile: 

Out of interest, is the test & assertion assertion code (`require('ava')`) decoupled enough from the runner that it could be used made to work in a browser context? It might be possible to keep the babel & parallelisation code node-only, but get the test and assertion code browser-compatible.

**Comment by sindresorhus** [comment]
@phuu The test runner `lib/runner.js` is decoupled and can be used independently, see its [test](https://github.com/sindresorhus/ava/blob/master/test/runner.js) and how it's used in [index.js](https://github.com/sindresorhus/ava/blob/master/index.js).

**Comment by jokeyrhyme** [comment]
I found karma-tap, but there may or may not be an issue with how we end up using it: https://github.com/tmcw/karma-tap/issues/10

**Comment by jamestalmage** [comment]
The following will be helpful when we do move forward with this:

https://code.google.com/p/chromium/issues/detail?id=485425

**Comment by BarryThePenguin** [comment]
Would AVA provide a solution for browser testing through it's own web server, browser loader and client, or is the plan to just support, and have examples/recipes, for running tests in a browser environment?

**Comment by ariporad** [comment]
@BqrryThePenguin: (Awesome username, btw). We would almost certainly
support running tests through karma as the preferred browser setup.

**Comment by jamestalmage** [comment]
It depends on how easy it would be to modify karma to run multiple files at once. But karma would definitely be the first thing we look at.

I would like to see it be a bit more than recipes. If nothing else, worst case a shareable karma baseconfig that you can use without modification if you do things "the AVA way"

**Comment by ariporad** [comment]
@jamestalmage: I haven't used karma in a while, but IIRC, karma just takes
an array of files.

**Comment by jokeyrhyme** [comment]
If you're happy with a simulated browser environment, we can probably already write ava tests with this today: https://github.com/assaf/zombie

**Comment by jamestalmage** [comment]
@jokerhyme if that actually works, might make for a good recipe.

**Comment by mattkrick** [comment]
Any thoughts on building browser compatibility against a selenium server, something like http://nightwatchjs.org/ or https://theintern.github.io/?

**Comment by bfred-it** [comment]
@jokeyrhyme I'm not sure about the value of _generic browser testing_, considering that actual browsers vary a lot. Just because something works there it doesn't mean it works in real-world browsers.

**Comment by jamestalmage** [comment]
@jokeyrhyme - The eventual goal is to build in real-world browser testing. If not directly using karma's browser launchers, then something very similar.

**Comment by hoschi** [comment]
@jamestalmage commented on 14 Jan
> It depends on how easy it would be to modify karma to run multiple files at once. But karma would definitely be the first thing we look at.

In a first version it would be great to have serial tests, which run in a real browser. For performant dev tests you can use JSDOM or generic fake browsers. Test runs in real browser can run in a CI environment, where performance doesn't matter.

**Comment by novemberborn** [comment]
Some quick thoughts given #887:

* It seems sensible for AVA to spawn a proxy server to serve tests, sources and other fixtures
  * Fixtures need to be served with correct mime types
  * Transpilation should happen in AVA, shipping `babel-register` to the browser seems fraught
  * We'd somehow need to bundle the required JavaScript, figure out compatibility with the various loaders, etc
* Certain use cases can be solved by printing a URL to access the proxy
  * We may need to be able to serve with HTTPS certificates
* Other use cases would involve AVA controlling browsers, e.g. Saucelabs, WebDriver, etc
  * What can we learn / use from Karma?
* AVA would have to collect code coverage
* Let's not get hung up on running test files concurrently
* However each test file still needs a unique environment
  * Do we sandbox in iframes?
  * Or reload between tests?
* I'd be happy for test results to be reported back to the terminal, and in the Dev Tools console, but not shown in the web page itself

**Comment by vdemedes** [comment]
So, here's my results and thoughts regarding browser support implementation (#887).

### Overview

To run tests in the browser, AVA provides `--browser` flag, which:

1. Precompiles and assembles test environment (AVA itself) and tests using browserify
2. Starts an HTTP server, index page of which runs tests and displays results

### Tests in browser

As all of you know, AVA runs test files in an individual, isolated environment (a new node process) with ability to configure concurrency. In the browser, AVA offers the exact same functionality by taking advantage of Web Workers. For each new test file, AVA spawns a new Worker. Communication between main thread and workers happens via standard methods provided by browser: `onmessage` and `postMessage`. Those are basically identical to `process.on('message')` and `process.send` in Node.js.

### Watch mode

Watch mode will be supported by leveraging SSE (Server-Sent Events) or WebSockets. AVA process, that's running in the console will watch for changes (same as now), only with `--browser` flag it will assemble a new bundle of tests and notify browser to download & run new tests.

### Babel

AVA's Babel-related functionality remains in place. Files are pre-compiled when building a bundle, so `babel` is not included in the bundle and served to a client.

### Under the hood

Above mentioned PR tries to re-use the more code as possible from AVA core. These new files needed to be added:

- `browser-index.js` - the same `index.js`, but uses Web Worker APIs
- `browser-fork.js` - the same `fork.js`, but spawns Web Workers instead of node processes
- `browser-test-worker.js` - the same `test-worker.js`, but does not use require hooks

There's also a `browser.js`, which serves as `cli.js`, but for the browser. It reads configuration, talks to API and listens for events from the AVA server.

**Comment by vdemedes** [comment]
We should also implement #709 to disable Babel, in case users want to test cross-browser compatibility. Not entirely, but only remove ES6/7 presets. Our plugins, like power-assert, should remain untouched.

**Comment by jokeyrhyme** [comment]
@vdemedes any thoughts about how we could accomplish DOM tests (or other things not available in Web Workers)?

**Comment by jamestalmage** [comment]
How far back does browser support for WebWorkers go? Especially IE. I think it's not very far.

Couldn't we accomplish much the same thing via iframes?

**Comment by cloudkite** [comment]
Web workers are only supported in IE11+ see http://caniuse.com/#feat=webworkers

Also you don't have access to a bunch of things in web workers such as the window or DOM in web workers so it would limit the usefulness of browser testing

**Comment by bfred-it** [comment]
Web Workers are not good enough for all situations; even iframes might not be, since `self !== top` and some tests might fail because of this discrepancy.

I think the best solutions would be using iframes, given its wide support of browsers and capabilities (DOM access, etc), **plus** a full mode, perhaps optional, that requires full window access, which if we cannot open multiple tabs "visible" it would mean that tests become serial. "Visible" is the keyword here, `requestAnimationFrame` among others wouldn't work in hidden tabs and things sometimes behave differently.

**Comment by vdemedes** [comment]
Yeah, I forgot about DOM in Web Workers. I think I need to switch that PR to iframes, unfortunately.

Although with iframes debugging of tests would be complicated, because we'd remove the iframe after tests finished and its output wouldn't be visible in the "main" page.

**Comment by bfred-it** [comment]
Are you talking about my reference to "visible"? As long as the tab is the foremost tab, every iframe in it will be "visible". Once the test in the iframe is done, it can be removed and the output/visibility is irrelevant, I'd think.

**Comment by jamestalmage** [comment]
There is a new flag in Chrome that causes `requestAnimationFrame` to work, even when the window isn't visible. I linked to it above, but the issue has since been restricted to collaborators (you can't even read it). I'm not sure what that means, but it's a flag we probably want to enable.

**Comment by jamestalmage** [comment]
I think we should try to figure out how to reuse karma wherever possible. Probably launchers. Maybe the browser to console relay.

Karma already allows concurrent testing across multiple browsers (though not concurrency within the test run for an individual browser).

We should try to reuse whatever we can.

> I'd be happy for test results to be reported back to the terminal, and in the Dev Tools console, but not shown in the web page itself

:+1:

**Comment by FredyC** [comment]
I am wondering if there is anything that I could do to help with this. 

Personally I wouldn't even mind if those tests would not run in parallel. If I delegate test running to some service like Saucelabs and it will run all tests in series for each browser, it should be more than fine. Going for hassle with web workers just seems like overkill.

**Comment by jamestalmage** [comment]
> Personally I wouldn't even mind if those tests would not run in parallel

I think that's a fine first goal, and if it gets us to browser support faster, so be it.

> Going for hassle with web workers just seems like overkill.

We've identified other problems with web workers (see above), so that's definitely off the table. 

That said, isolation is way more important than parallel execution. The two best solutions for isolation (WebWorkers and iframes) also facilitate parallel execution, so I'm not sure we need to punt on either for long.

I think iframes should work OK. We may need to come up with something to allow for apps that care about `window.top`, but we can punt on that for now.

> I am wondering if there is anything that I could do to help with this

Reviewing and commenting on #887 is a good start. If you've got alternate ideas, proof-of-concept PR's wouldn't hurt.

**Comment by FredyC** [comment]
> I think iframes should work OK. We may need to come up with something to allow for apps that care about window.top, but we can punt on that for now.

Sounds to me like trying to cover for some eventuality that is actually rarely needed. Or at least I don't have any experience when I would need to rely on `window.top`. In my case it's merely about having ability to run library unit tests in browser environment to ensure there isn't some incompatibility. As much as its nice to have tests written in same way, I wouldn't probably choose Ava for writing/running E2E tests as there are surely some other more mature tools for that.

I will have a look at that PR, thanks.

**Comment by bfred-it** [comment]
Yeah, surely iframe issues are rare, but if you never know what hacks polyfills and such have to rely on. Also what happens when two iframes try to interact with the History API? That's another one

**Comment by jamestalmage** [comment]
Hmm. Maybe we should start with running test files one at a time, top-level, launching a new browser/tab for each one. That will be really slow, but it's probably the easiest to accomplish. Then we can work towards parallelism.

**Comment by andrewmclagan** [comment]
That sounds like a splendid idea.

**Comment by novemberborn** [comment]
I'm curious what kind of browser support people are looking for.

* Do you want to see test results in the browser? In the CLI? Both?
* Do you start browser tests manually by navigating to a URL, or should AVA launch your browsers?
* How would you run browser tests in CI? Through SauceLabs / BrowserStack / others, or local VMs?
* Does your code need to be compiled specifically to be tested in a browser environment?

Please add anything else that seems relevant to the discussion.

**Comment by bfred-it** [comment]
I'd use this on Travis+SauceLabs to verify real compatibility with browsers, rather than "compatibility with jsdom."  _My_ issue is I have no experience in setting this up, so documentation specific to AVA would be great, perhaps that's the only thing missing.

**Comment by hoschi** [comment]
I give it a try with my current setup and use cases. On local (developer) machine tests get executed in jsdom environment for best performance. On CI server these tests should run in a real browser environment. If one of these tests fail, I want to be able to debug it with browsers dev tools.

> Do you want to see test results in the browser? In the CLI? Both?
CLI or file to save it for each build.

> Do you start browser tests manually by navigating to a URL, or should AVA launch your browsers?
Both is needed to run a specific test in a specific browser and automate the process in a whole?

> How would you run browser tests in CI? Through SauceLabs / BrowserStack / others, or local VMs?
Not used these kind of tools for a long time, no real opinion here, but BroswerStack sounds good.

> Does your code need to be compiled specifically to be tested in a browser environment?
As this are unit tests, I think dev compiled is ok. Productivity compiled code is for integration tests, I would say. Test code/setup should be configurable like the "--require" flag already does, at the moment I setup jsdom with this flag.

**Comment by gajus** [comment]
Can the title of this issue be set to "Browser support". The "Browserify" part makes it sound of-topic.

**Comment by sindresorhus** [comment]
Done

**Comment by elaijuh** [comment]
I am finding a way to do angular 2 unit test. All I could find is karma + jasmine + browser launcher. What I expect is I could use AVA to tackle it, so this issue looks like working on providing a browser environment for client side code.

I have tried `browser-env` with no luck working together with `zone.js`, client side support is a solid thing to me.

**Comment by mAAdhaTTah** [comment]
I'm currently playing around with using AVA to unit test a framework I'm working on (wooo Observables!), but without being able to run the unit tests themselves in the browser, it's a non-starter. I'm definitely interested in helping out; I played around with `karma-ava` but couldn't get it working.

Is there any movement on this, or anything I can look at or help out with?

**Comment by novemberborn** [comment]
@mAAdhaTTah we're still evolving AVA's internals too much to really commit to browser support, sorry.

**Comment by wearhere** [comment]
> I'm curious what kind of browser support people are looking for.

To be really specific, the ideal outcome for Mixmax would be for Ava to work with the [Karma](https://karma-runner.github.io/1.0/index.html) test runner. What Karma lets you do is take the same test files that you would run in a Node environment, and run them using real browsers or devices or PhantomJS or headless Chrome instances. Karma launches the browser, and logs results to the CLI (at least how we use it); and as far as CI goes, headless Chrome works just fine, without special configuration, on Travis. The tests do need to be compiled for the browser i.e. bundle any dependencies (including Node modules) into a single IIFE or load them as globals.

We see that there is a Karma plugin for Ava [here](https://github.com/avajs/karma-ava) but it says "alpha level" currently, not sure if that's blocked on anything in this project.

(Want to try out Karma? See installation and configuration instructions [here](https://github.com/karma-runner/karma#installation), though we can't vouch for those, we use/maintain https://github.com/mixmaxhq/erik i.e. currently use Jasmine for running unit tests in the browser, Ava only for server tests—though it would be nice to use Ava everywhere!)

**Comment by novemberborn** [comment]
Hey all, it's hopefully become apparent since this issue was first opened that browser support is not a priority for AVA. I am therefore closing this issue.

We are interested in letting you use AVA to drive browser tests, through Karma or other solutions. If you've got experience with this we'd love to see updates to the [browser testing](https://github.com/avajs/ava/blob/master/docs/recipes/browser-testing.md) recipe, or indeed new recipes.

If you want to help out integrating browser tests by building on top of AVA then please [get in touch](https://gitter.im/avajs/ava). We don't have the resources to spearhead these efforts ourselves, but we'd love to include mature solutions in our GitHub and npm organizations, or promote them in recipes. Forking [`karma-ava`](https://github.com/avajs/karma-ava) might be a good start.

Thank you all for your interest in AVA and participation in this issue ❤️

**Comment by bj0** [comment]
sorry to bump an old issue, but I just tried using ava for the first time to test some client side js that uses Blob and SubtleCrypto apis, but they are not defined i node (I guess?  I've never used node...).  

Is the only way to use these APIs to run in browser?

**Comment by novemberborn** [comment]
@bj0 you'll have to stub those APIs in Node.js. Have a look at our [browser testing](https://github.com/avajs/ava/blob/master/docs/recipes/browser-testing.md) recipe — the libraries mentioned there may already implement these APIs. (Or not, I haven't actually checked for you, sorry.)

**Comment by fabiospampinato** [comment]
Hello there, for anybody interested in this I'm kind of rewriting ava for browser support: https://github.com/fabiospampinato/fava, the thing currently has **many** drawbacks compared to ava, but it's designed to work in the browser, so if you need that maybe give it a try, if you find issues with it I'll be happy to address them.

It's not quite the same as just running ava itself in the browser, but the APIs are largely the same so you can switch back to ava once that's implemented here if you like.

**Comment by nicholascelestin** [comment]
Hello There,

For anyone interested in running ava tests in a browser, I've got a way to do that documented in this PR: #3086  . Basically, you run ava in a node process like usual, and a macro delegates execution of the test code to a headless Puppeteer browser.
